oeps/oep-0030-arch-pii-markup-and-auditing.rst

```diff
@@ -0,0 +1,269 @@
+
+Example 2::
+
+    # ..pii:: Learner email is sent to Segment in the following line and will be associated with analytics data. Retired: In Segment retirement step in Tubular.
+
+The goal of this is to allow creation of `Documentation Tooling`_ which will automatically create documentation listing all of the known locations of PII in each repository.
+
+**Javascript**
+
+When adding in Javascript that results in storage of PII to a location that is not covered by other annotations (ex. Segment), annotations should be added to the location(s) in script where the data is being sent. The annotations should take the same form as in Python as Sphinx can also operate on Javascript for documentation.
+
+Example 1::
+
+    % if settings.LMS_SEGMENT_KEY:
+        <!-- begin segment footer -->
+        <!-- .. pii:: The user is identified to Segment by username and email here, analytics data will be associated with this user based on their browsing. See Segment documentation for details. Retired: In Segment retirement step in Tubular -->
+        <script type="text/javascript">
+        % if user.is_authenticated:
+            ...
+
+Example 2::
+
+    <script type="text/javascript">
+    // .. pii:: The user's address email address is sent to the billing provider here. Retired: This information is not retired as it is necessary to keep for legal and financial reporting reasons.
+    </script>
+
+
+Example 3::
+
+    <script type="text/javascript">
+    /* .. pii:: Updates the user's email address with our email marketing provider. Retired: In the Tubular RETIRE_EMAIL step. */
+    </script>
+
+**Other Cases**
+
+It is likely that other use cases will come up that encompass new languages and storage. In those cases attempts should be made to make those cases match the designs laid out here for making PII locations auditable at the repository level and this OEP should be updated to include best practices for the new case.
+
+Enforcement Tooling
+-------------------
+A tool will be created and integrated into the Open edX test / build systems that will examine all Django models in the project and ensure that they have PII annotations. It is acknowledged that this tool will not handle all cases where PII is stored, but represents an effort to enforce best practices on the majority of places where PII is stored in the Open edX ecosystem.
+
+This tool may take the form of a test run inside the project which looks at all installed apps and their models for docstrings containing PII. Given that this list will contain many third party packages we will also need to maintain a list of packages and models that will not be checked, as well as a list of PII stored in those packages / models. These lists will need to be hand maintained by the developers adding or modifying packages. This mechanism will also allow the rollout of the annotations to take place over time.
+
```

Comment:
For our own open edX code, doing static analysis makes sense.  For 3rd party database models/tables, is there a more automated approach that we can use at runtime?  Maintaining the information by hand - branching down the dependency hierarchy - every time we upgrade a library - will be costly.  If there is no alternative, then so be it.  But I'm wondering if there's tooling that can be a backup in case we err.